test/src/test/java/hudson/EnvVarsTest.java

```diff
@@ -0,0 +1,143 @@
+        resetEnvVarsCache();
+        
+        FreeStyleProject p = j.createFreeStyleProject();
+        CaptureEnvironmentBuilder ceb = new CaptureEnvironmentBuilder();
+        p.getBuildersList().add(ceb);
+        j.assertBuildStatusSuccess(p.scheduleBuild2(0, new Cause.UserCause(), Arrays.asList(
+                new ParametersAction(
+                        new StringParameterValue("var1", "value1"),
+                        new StringParameterValue("Var1", "value2")
+                )
+        )));
+        assertEquals("value1", ceb.getEnvVars().get("var1"));
+        assertEquals("value2", ceb.getEnvVars().get("Var1"));
+        assertNull(ceb.getEnvVars().get("VAR1"));
+    }
+    
+    @Test
+    public void testCaseSensitiveOnSlave() throws Exception {
+        assertNull(System.getProperty(EnvVars.PROP_CASE_INSENSITIVE));
+        resetEnvVarsCache();
+        
+        DumbSlave slave = j.createOnlineSlave();
+        FreeStyleProject p = j.createFreeStyleProject();
+        p.setAssignedNode(slave);
+        CaptureEnvironmentBuilder ceb = new CaptureEnvironmentBuilder();
+        p.getBuildersList().add(ceb);
+        j.assertBuildStatusSuccess(p.scheduleBuild2(0, new Cause.UserCause(), Arrays.asList(
+                new ParametersAction(
+                        new StringParameterValue("var1", "value1"),
+                        new StringParameterValue("Var1", "value2")
+                )
+        )));
+        assertEquals("value1", ceb.getEnvVars().get("var1"));
+        assertEquals("value2", ceb.getEnvVars().get("Var1"));
+        assertNull(ceb.getEnvVars().get("VAR1"));
+    }
+    
+    @Test
+    public void testCaseInsensitiveOnMaster() throws Exception {
+        assertNull(System.getProperty(EnvVars.PROP_CASE_INSENSITIVE));
+        try {
+            System.setProperty(EnvVars.PROP_CASE_INSENSITIVE, "true");
+            resetEnvVarsCache();
+            
+            FreeStyleProject p = j.createFreeStyleProject();
+            CaptureEnvironmentBuilder ceb = new CaptureEnvironmentBuilder();
+            p.getBuildersList().add(ceb);
+            j.assertBuildStatusSuccess(p.scheduleBuild2(0, new Cause.UserCause(), Arrays.asList(
+                    new ParametersAction(
+                            new StringParameterValue("var1", "value1"),
+                            new StringParameterValue("Var1", "value2")
+                    )
+            )));
+            String expected = ceb.getEnvVars().get("var1");
+            assertEquals(expected, ceb.getEnvVars().get("Var1"));
+            assertEquals(expected, ceb.getEnvVars().get("VAR1"));
+        } finally {
+            System.getProperties().remove(EnvVars.PROP_CASE_INSENSITIVE);
+        }
+    }
+    
+    @Test
+    public void testCaseInsensitiveOnSlave() throws Exception {
+        assertNull(System.getProperty(EnvVars.PROP_CASE_INSENSITIVE));
+        try {
+            System.setProperty(EnvVars.PROP_CASE_INSENSITIVE, "true");
```

Comment:
Well, in the case of Jenkins tests, the harness allows multiple JVMs to be forked and run concurrently, but only one test case runs in a given JVM at a time. Otherwise we would have all kinds of problems, since you cannot currently run multiple copies of Jenkins in one JVM. So plenty of tests use `try`-`finally` blocks to temporarily set some static state, I am afraid.